### DISCUSSION TITLE
using instances() across modules

### DISCUSSION THREAD

**Discussion by warmuuh** [kcl-lang/kcl#1588] [Q&A] [answered]
I have a "common" module which should also trigger `manifests.yaml_stream` on all instances of `A`.
```
manifests.yaml_stream(sum([render(a) for a in common.A.instances()], []))
```
then, i use that common module in another module `client` which simply defines an instance of `A`.
the only output i get is `null`.

moving this call into the `client` module works. is this by design? so do i need to collect the instances always in the main module? (this would mean to always add an extra line, such as `common.render(common.A.instances())` in each of the users of `common`.

**Comment by Peefy** [comment, marked as answer]
Have you tried this argument? 

![image](https://github.com/user-attachments/assets/ec85ee2d-9bb5-4222-ba0c-fa50213479e6)

See the change list in KCL v0.9.x https://www.kcl-lang.io/blog/2024-07-05-kcl-0.9.0-release#language

Because obtaining instances across packages is a side effect function, by default only schema instances from the main package will be obtained.

**Reply by warmuuh** [reply]
i already played around with that parameter, but it didnt work... should i create a minimal example?

ok, i think, i found the problem...
it has to do with the imports:

in the client, if i import the file, not the package, defining the schema (such as `import module.a` for the `module/a.k`  file in common module defining schema A) and use it (`a.A`), all is fine.
if i import the package (`import module`) and use it (`module.A`), all is good except that `a.A.instances()` is empty, if called from the common module.

**Reply by Peefy** [reply]
Yes, we need to keep the module package path same with `module` instead of `module.a`
